File: tripper-sdk/maneuvers.py
"""Google Navigation SDK maneuver → Tripper icon bytes."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, Tuple

MAN_HIGHWAY = 0x60


@dataclass(frozen=True)
class GoogleManeuver:
    name: str
    byte5: int
    byte6: int
    display_name: str


GOOGLE_MANEUVERS: Tuple[GoogleManeuver, ...] = (
    GoogleManeuver("DEPART", 0x00, 0x40, "Depart"),
    GoogleManeuver("STRAIGHT", 0x00, 0x40, "Straight"),
    GoogleManeuver("DESTINATION_LEFT", 0x10, 0x01, "Destination left"),
    GoogleManeuver("DESTINATION_RIGHT", 0x20, 0x01, "Destination right"),
    GoogleManeuver("TURN_LEFT", 0x10, 0x40, "Turn left"),
    GoogleManeuver("TURN_RIGHT", 0x20, 0x40, "Turn right"),
    GoogleManeuver("TURN_KEEP_LEFT", 0x10, 0x40, "Keep left"),
    GoogleManeuver("TURN_KEEP_RIGHT", 0x20, 0x40, "Keep right"),
    GoogleManeuver("TURN_SLIGHT_LEFT", 0x10, 0x40, "Slight left"),
    GoogleManeuver("TURN_SLIGHT_RIGHT", 0x20, 0x40, "Slight right"),
    GoogleManeuver("TURN_SHARP_LEFT", 0x30, 0x40, "Sharp left"),
    GoogleManeuver("TURN_SHARP_RIGHT", 0x30, 0x40, "Sharp right"),
    GoogleManeuver("TURN_U_TURN_CLOCKWISE", 0x50, 0x40, "U-turn CW"),
    GoogleManeuver("TURN_U_TURN_COUNTERCLOCKWISE", 0x50, 0x40, "U-turn CCW"),
    GoogleManeuver("FORK_LEFT", 0x10, 0x40, "Fork left"),
    GoogleManeuver("FORK_RIGHT", 0x20, 0x40, "Fork right"),
    GoogleManeuver("KEEP_LEFT", 0x10, 0x40, "Keep left"),
    GoogleManeuver("KEEP_RIGHT", 0x20, 0x40, "Keep right"),
    GoogleManeuver("MERGE_LEFT", 0x10, 0x40, "Merge left"),
    GoogleManeuver("MERGE_RIGHT", 0x20, 0x40, "Merge right"),
    GoogleManeuver("ON_RAMP_LEFT", MAN_HIGHWAY, 0x40, "On-ramp left"),
    GoogleManeuver("ON_RAMP_RIGHT", MAN_HIGHWAY, 0x40, "On-ramp right"),
    GoogleManeuver("OFF_RAMP_LEFT", MAN_HIGHWAY, 0x40, "Off-ramp left"),
    GoogleManeuver("OFF_RAMP_RIGHT", MAN_HIGHWAY, 0x40, "Off-ramp right"),
    GoogleManeuver("ROUNDABOUT_LEFT_CLOCKWISE", 0x40, 0x10, "Roundabout left"),
    GoogleManeuver("ROUNDABOUT_RIGHT_CLOCKWISE", 0x40, 0x20, "Roundabout right"),
    GoogleManeuver("ROUNDABOUT_STRAIGHT_CLOCKWISE", 0x40, 0x40, "Roundabout straight"),
    GoogleManeuver("ROUNDABOUT_EXIT_CLOCKWISE", 0x40, MAN_HIGHWAY, "Roundabout exit"),
    GoogleManeuver("ROUNDABOUT_U_TURN_CLOCKWISE", 0x40, 0x50, "Roundabout U-turn"),
    GoogleManeuver("ROUNDABOUT_LEFT_COUNTERCLOCKWISE", 0x40, 0x10, "Roundabout left CCW"),
    GoogleManeuver("ROUNDABOUT_RIGHT_COUNTERCLOCKWISE", 0x40, 0x20, "Roundabout right CCW"),
    GoogleManeuver("ROUNDABOUT_EXIT_COUNTERCLOCKWISE", 0x40, MAN_HIGHWAY, "Roundabout exit CCW"),
)

_BY_NAME: Dict[str, GoogleManeuver] = {m.name: m for m in GOOGLE_MANEUVERS}


def get_google_maneuver(name: str) -> GoogleManeuver | None:
    return _BY_NAME.get(name)


def nav_maneuver_to_byte(maneuver_id: int, roundabout_exit: int = 0) -> int:
    """Map Google Navigation SDK maneuver id (NavManeuver.toByte)."""
    table = {
        0: 0x00, 1: 0x00, 2: 0x10, 3: 0x20, 4: 0x30, 5: 0x40, 6: 0x50, 7: 0x60,
        8: 0x21, 9: 0x22, 10: 0x1F, 11: 0x20, 12: 0x1D, 13: 0x1E, 14: 0x1B, 15: 0x17,
        16: 0x16, 17: 0x15, 18: 0x14, 19: 0x3C, 20: 0x2D, 21: 0x2E, 22: 0x2B, 23: 0x31,
        24: 0x30, 25: 0x2F, 26: 0x18, 27: 0x19, 28: 0x1A, 29: 0x3D, 30: 0x3E, 31: 0x3F,
        32: 0x09, 33: 0x0A, 34: 0x08, 36: 0x1B, 37: 0x1A, 38: 0x19, 39: 0x18, 40: 0x17,
        41: 0x16, 42: 0x15,
    }
    if maneuver_id == 35:
        return (0x1C + roundabout_exit) if roundabout_exit > 0 else 0xFF
    return table.get(maneuver_id, 0xFF)

File: src/bluetooth/tripper/maneuvers.ts
import { MAN_HIGHWAY } from "./constants";

export interface GoogleManeuverDef {
  name: string;
  byte5: number;
  byte6: number;
  displayName: string;
}

/** Google Navigation SDK maneuver → Tripper icon bytes (from TripperProtocol.GoogleManeuver). */
export const GOOGLE_MANEUVERS: GoogleManeuverDef[] = [
  { name: "DEPART", byte5: 0x00, byte6: 0x40, displayName: "Depart" },
  { name: "STRAIGHT", byte5: 0x00, byte6: 0x40, displayName: "Straight" },
  { name: "DESTINATION_LEFT", byte5: 0x10, byte6: 0x01, displayName: "Destination left" },
  { name: "DESTINATION_RIGHT", byte5: 0x20, byte6: 0x01, displayName: "Destination right" },
  { name: "TURN_LEFT", byte5: 0x10, byte6: 0x40, displayName: "Turn left" },
  { name: "TURN_RIGHT", byte5: 0x20, byte6: 0x40, displayName: "Turn right" },
  { name: "TURN_KEEP_LEFT", byte5: 0x10, byte6: 0x40, displayName: "Keep left" },
  { name: "TURN_KEEP_RIGHT", byte5: 0x20, byte6: 0x40, displayName: "Keep right" },
  { name: "TURN_SLIGHT_LEFT", byte5: 0x10, byte6: 0x40, displayName: "Slight left" },
  { name: "TURN_SLIGHT_RIGHT", byte5: 0x20, byte6: 0x40, displayName: "Slight right" },
  { name: "TURN_SHARP_LEFT", byte5: 0x30, byte6: 0x40, displayName: "Sharp left" },
  { name: "TURN_SHARP_RIGHT", byte5: 0x30, byte6: 0x40, displayName: "Sharp right" },
  { name: "TURN_U_TURN_CLOCKWISE", byte5: 0x50, byte6: 0x40, displayName: "U-turn CW" },
  { name: "TURN_U_TURN_COUNTERCLOCKWISE", byte5: 0x50, byte6: 0x40, displayName: "U-turn CCW" },
  { name: "FORK_LEFT", byte5: 0x10, byte6: 0x40, displayName: "Fork left" },
  { name: "FORK_RIGHT", byte5: 0x20, byte6: 0x40, displayName: "Fork right" },
  { name: "KEEP_LEFT", byte5: 0x10, byte6: 0x40, displayName: "Keep left" },
  { name: "KEEP_RIGHT", byte5: 0x20, byte6: 0x40, displayName: "Keep right" },
  { name: "MERGE_LEFT", byte5: 0x10, byte6: 0x40, displayName: "Merge left" },
  { name: "MERGE_RIGHT", byte5: 0x20, byte6: 0x40, displayName: "Merge right" },
  { name: "ON_RAMP_LEFT", byte5: MAN_HIGHWAY, byte6: 0x40, displayName: "On-ramp left" },
  { name: "ON_RAMP_RIGHT", byte5: MAN_HIGHWAY, byte6: 0x40, displayName: "On-ramp right" },
  { name: "OFF_RAMP_LEFT", byte5: MAN_HIGHWAY, byte6: 0x40, displayName: "Off-ramp left" },
  { name: "OFF_RAMP_RIGHT", byte5: MAN_HIGHWAY, byte6: 0x40, displayName: "Off-ramp right" },
  { name: "ROUNDABOUT_LEFT_CLOCKWISE", byte5: 0x40, byte6: 0x10, displayName: "Roundabout left" },
  { name: "ROUNDABOUT_RIGHT_CLOCKWISE", byte5: 0x40, byte6: 0x20, displayName: "Roundabout right" },
  { name: "ROUNDABOUT_STRAIGHT_CLOCKWISE", byte5: 0x40, byte6: 0x40, displayName: "Roundabout straight" },
  { name: "ROUNDABOUT_EXIT_CLOCKWISE", byte5: 0x40, byte6: MAN_HIGHWAY, displayName: "Roundabout exit" },
  { name: "ROUNDABOUT_U_TURN_CLOCKWISE", byte5: 0x40, byte6: 0x50, displayName: "Roundabout U-turn" },
  { name: "ROUNDABOUT_LEFT_COUNTERCLOCKWISE", byte5: 0x40, byte6: 0x10, displayName: "Roundabout left CCW" },
  { name: "ROUNDABOUT_RIGHT_COUNTERCLOCKWISE", byte5: 0x40, byte6: 0x20, displayName: "Roundabout right CCW" },
  { name: "ROUNDABOUT_EXIT_COUNTERCLOCKWISE", byte5: 0x40, byte6: MAN_HIGHWAY, displayName: "Roundabout exit CCW" },
];

const googleManeuverByName = new Map(GOOGLE_MANEUVERS.map((m) => [m.name, m]));

export function getGoogleManeuver(name: string): GoogleManeuverDef | undefined {
  return googleManeuverByName.get(name);
}

/**
 * Map Google Navigation SDK maneuver id to Tripper maneuver byte (NavManeuver.toByte).
 * Returns 0xFF for unknown maneuvers.
 */
export function navManeuverToByte(maneuverId: number, roundaboutExit = 0): number {
  switch (maneuverId) {
    case 0:
    case 1:
      return 0x00;
    case 2:
      return 0x10;
    case 3:
      return 0x20;
    case 4:
      return 0x30;
    case 5:
      return 0x40;
    case 6:
      return 0x50;
    case 7:
      return 0x60;
    case 8:
      return 0x21;
    case 9:
      return 0x22;
    case 10:
      return 0x1f;
    case 11:
      return 0x20;
    case 12:
      return 0x1d;
    case 13:
      return 0x1e;
    case 14:
      return 0x1b;
    case 15:
      return 0x17;
    case 16:
      return 0x16;
    case 17:
      return 0x15;
    case 18:
      return 0x14;
    case 19:
      return 0x3c;
    case 20:
      return 0x2d;
    case 21:
      return 0x2e;
    case 22:
      return 0x2b;
    case 23:
      return 0x31;
    case 24:
      return 0x30;
    case 25:
      return 0x2f;
    case 26:
      return 0x18;
    case 27:
      return 0x19;
    case 28:
      return 0x1a;
    case 29:
      return 0x3d;
    case 30:
      return 0x3e;
    case 31:
      return 0x3f;
    case 32:
      return 0x09;
    case 33:
      return 0x0a;
    case 34:
      return 0x08;
    case 35:
      return roundaboutExit > 0 ? 0x1c + roundaboutExit : 0xff;
    case 36:
      return 0x1b;
    case 37:
      return 0x1a;
    case 38:
      return 0x19;
    case 39:
      return 0x18;
    case 40:
      return 0x17;
    case 41:
      return 0x16;
    case 42:
      return 0x15;
    default:
      return 0xff;
  }
}
